rfcs/0000-local-overlay-store.md

```diff
@@ -0,0 +1,388 @@
+This is because when the `local-overlay` is remounted, it doesn't know how the lower store may have changed, and when the lower store is added to, any upper store directories / DBs are not in general visible either.
+We can have a "fsck" operation however that manually scans for missing / duplicated objects.
+
+## Read-only `local` Store
+
+In order to facilitate using `local-overlay` where the lower store is entirely read only (read only SQLite files too, not just store directory), it is useful to also implement a new "read-only" setting on the `local` store.
+The main thing this does is use SQLite's [immutable mode](https://www.sqlite.org/c3ref/open.html).
+
+This is a separate feature;
+it is perfectly possible to implement the `local-overlay` without this or vice-versa.
+But for maximum usability, we want to do both.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Because the `local-overlay` store is a completely separate store implementation, the interactions with the rest of Nix are fairly minimal and well-defined.
+In particular, users of other stores and not the `local-overlay` store will not be impacted at all.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Read-only local store is delicate
+
+SQLite's immutuable mode doesn't mean "we promise not to change the database".
+It relies on the database not just being *logically* immutable (the meaning doesn't change), but *physically* immutable (no bytes of the on-disk files change).
+This is because it is forgoing synchronization altogether and thus relying that nothing can be rearranged e.g. in the middle of a query (invalidating in-progress traversals of data structures, etc.).
+This means there is no hope of, say "append only" mode where "publishers" only add new store objects to a local store, while read-only mode "subscribers" are able to deal with other store objects just fine.
+
+This is an inconvenience for rolling out new versions of the lower store, but not a show stopper.
+One solution is "multi version concurrency control" where consumers get an immutable snapshot of the lower store for the duration of each login.
+New snapshots can only be gotten when consumers log in again, and old snapshots can only be retired once every consumer viewing them logs out.
+
+## `local-overlay` lacks normal form for the database
+
+A slight drawback with the architecture is a lack of a normal form.
+A store object in the lower store may or may not have a DB entry in the `overlay-local` store.
+
+> In the store object state diagram diagram, this is represented by the fact that there are two green nodes, instead of just one like the one blue node.
+
+This introduces some flexibility in the system: the same "logical" layered store can be represented in multiple different "physical" configurations.
+This isn't a problem *per se*, but does mean there is a bit more complexity to consider during testing and system administration.
```

Comment:
> It would be good to have some elaborations on what the complexity increase will be.
Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

Part of my vagueness is that I suspect there is some ambiguity I haven't thought of.

> Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

That at least I don't think can happen.

Reply:
> If the system-admin does a mistake on remounting, and mounts a different lower layer, is there any way to detect it during init()?
> or things will blow up in runtime?

The "fsck" thing would catch it, but I fear it will be slow enough that it shouldn't be run unconditionally.

Reply:
> Could a situation happen, where the user could end up "using" from the lower-layer, without having an entry in overlay db?

Here I was trying to refer to a situation where a user does a wrong thing, like copy a dir from one system to another, which refer to some nix store paths.
Doing such a thing on other nix stores would cause it to fail, but it would likely "work" in this case. Or it might fail in unexpected manner, if these copied nix-store links are used to do some nix commands by the user.
This isn't a big problem, but it still is a case of non "normal" behaviour.

> The "fsck" thing would catch it, but I fear it will be slow enough that it shouldn't be run unconditionally.

This isn't an important issue at the moment, could be ignored..